pxd_fastpath.c

```diff
@@ -455,6 +410,56 @@ static void __pxd_cleanup_block_io(struct pxd_io_tracker *head)
 	bio_put(&head->clone);
 }
 
+static void pxd_io_failover(struct work_struct *ws)
+{
+	struct pxd_io_tracker *head = container_of(ws, struct pxd_io_tracker, wi);
+	struct pxd_device *pxd_dev = head->pxd_dev;
+	bool cleanup = false;
+
+	BUG_ON(head->magic != PXD_IOT_MAGIC);
+	BUG_ON(pxd_dev->magic != PXD_DEV_MAGIC);
+
+	spin_lock(&pxd_dev->fp.fail_lock);
+	if (pxd_dev->fp.fastpath &&
+		pxd_dev->fp.active_failover == PXD_FP_FAILOVER_ACTIVE) {
+		pxd_dev->fp.active_failover = PXD_FP_FAILOVER_COMPLETE;
+		cleanup = true;
+	}
+	spin_unlock(&pxd_dev->fp.fail_lock);
+
+	if (cleanup) {
+		disableFastPath(pxd_dev, true);
+		pxd_resume_io(pxd_dev);
+	}
+
+	if (!pxd_dev->connected || pxd_dev->removing) {
+		printk_ratelimited(KERN_ERR"%s: pxd%llu: px is disconnected, failing IO.\n", __func__, pxd_dev->dev_id);
+		BIO_ENDIO(head->orig, -ENXIO);
+	} else {
+		// switch to native path
+		printk_ratelimited(KERN_ERR"%s: pxd%llu: resuming IO in native path.\n", __func__, pxd_dev->dev_id);
+		atomic_inc(&pxd_dev->fp.nslowPath);
+		pxd_make_request_slowpath(pxd_dev->disk->queue, head->orig);
+	}
+
+	__pxd_cleanup_block_io(head);
+	pxd_check_q_decongested(pxd_dev);
+}
+
+static void pxd_failover_initiate(struct pxd_device *pxd_dev, struct pxd_io_tracker *head)
```

Comment:
have to do this from non interrupt schedulable context, hence punting through a work queue.